Composer updates

lauramthomas opened the pull request:
<!-- Thanks for sending us a PR to improve this project! If you are adding a
feature or fixing a bug, and this needs more documentation, please add it to your PR.

**Thank you for contributing to the `sauce-docs` project!**
**A well described pull request helps maintainers quickly review and merge your change**

Before submitting your PR, please check our [contributing](https://github.com/saucelabs/sauce-docs/blob/main/docs/contributing.md)
guidelines, applied for this repository. Avoid large PRs, help reviewers by making them as simple
and short as possible. -->


<!--- Provide a general summary of your changes in the Title above -->

### Description
I moved the how-to content to the Composer section, and added/updated screenshots.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation fix (typos, incorrect content, missing content, etc.)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/saucelabs/sauce-docs/blob/main/docs/contributing.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

spider-sauce reviewed (approved):
LGTM 👍

LoryApiFortress commented:
@lauramthomas do not proceed with the merge, I need time to review

LoryApiFortress commented on docs/api-testing/composer.md:
```diff
@@ -132,12 +651,10 @@ There are two types of input data sets you can use:
-
-
-
+This button toggles back and forth between the Input Set and Unit views.
```
I would rather say "Activates/Shows the Unit" b/c is not a "toggle" or a button that you can activate/deactivate. Same for "Input"

lauramthomas replied:
@LoryApiFortress Changed to "These buttons switch between the Input Set and Unit views."

LoryApiFortress reviewed (commented):
see all my notes

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
-### Add Test Components
-The next of set steps is to add the necessary test components. When combined, they act as our test logic. You can choose from [I/O (Request) Components](/api-testing/composer/io-components), [Assertion Components](/api-testing/composer/assertion-components/), and [Logical Components](/api-testing/composer/logical-components/).
+2. On the **Projects** page:
+  * If the project has no tests yet, in the **Write your own test** box, click **Use Composer**.
```
Fixed

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
-13. Edit the following details:
-    * **Expression**: `payload.downloads` - This expression checks for the field `"downloads"` in the `json` response body.<br/><img src={useBaseUrl('img/api-fortress/2021/01/assertDetails.png')} alt="Assert Exists Details"/>
-    * Leave the rest of the fields blank and click **Confirm changes** (checkmark icon).
+5. On the **Compose** tab, make sure the **Input** and **Visual** views are toggled on.
```
@LoryApiFortress In your original doc, you wrote "On the Compose tab, make sure the Input and Visual views are toggled on." As I said before, I don't want to compromise the accuracy of the content. But if it is incorrect, I will remove that sentence.

If by "input" you mean the screenshot doesn't show any details about what the user should enter/include as part of the procedure, this step shows the user where the toggles are, not what to do or include on the page.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  <img src={useBaseUrl('/img/api-testing/get-request-nav.png')} alt="Navigating to the GET request window" width="400"/>
+
+  3. In the **GET request** window, in the **Url** field, enter **ht<span>tps://</span>api.us-west-1.saucelabs.com/rest/v1/public/tunnels/info/versions**.
```
@LoryApiFortress I don't see any quotation marks. What do you mean exactly?

lauramthomas replied:
@LoryApiFortress The <span> tag is used to remove the auto-generated hyperlink in a url. It does not show up in the output.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  3. In the **GET request** window, in the **Url** field, enter **ht<span>tps://</span>api.us-west-1.saucelabs.com/rest/v1/public/tunnels/info/versions**.
+
+  This URL will return a json response body.
```
@LoryApiFortress Just FYI, in your original document you wrote "This URL will return a json response body."

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  <img src={useBaseUrl('/img/api-testing/assert-exists-nav.png')} alt="Navigating to the Assert exists window"/>
+
+3. In the **Assert exists** window, in the **Expression** field, enter `payload.downloads`. This expression checks for the **downloads** field in the json response body.
```
@LoryApiFortress In your original document you wrote "In the Assert exists window, in the Expression field, enter payload.downloads. This expression checks for the downloads field in the json response body." and the screenshot showed the same info as mine does.

Note: I DID inadvertently enter "payload.download" in my screenshot, which I will fix.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+4. Leave the rest of the fields blank and click the checkmark to confirm the changes.
+
+  <img src={useBaseUrl('/img/api-testing/assert-exists-window.png')} alt="Confirm changes"/>
```
@LoryApiFortress Fixed (see my last comment)

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+5. The result should look like the following:
+
+  <img src={useBaseUrl('/img/api-testing/assert-exists-final.png')} alt="What the Assert request should look like" width="500"/>
+
```
@LoryApiFortress This is an example of something I need you to at least get started. I don't know enough about API Testing yet to write an accurate example.

lauramthomas replied:
@LoryApiFortress Thank you. I added this example to the docs.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+For more information, see [Assertion Test Components](/api-testing/composer/assertion-components/).
+
+### Run the Test
+In the Composer, click the **Save** icon and then click **Run**.
```
@LoryApiFortress In your original doc you wrote "Before you run the test, select the Save icon," but I will remove that step.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+All test runs appear to the right of the Composer, under the test details and environment sections.
+
+<img src={useBaseUrl('/img/api-testing/test-runs.png')} alt="Test Runs in the Composer" width="350"/>
```
@LoryApiFortress The screenshot and the step show the user where the test runs are displayed, not the results of a specific test.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
 
-:::tip
-For detailed information, see [Assertion Components](/api-testing/composer/assertion-components/).
+<img src={useBaseUrl('/img/api-testing/input-visual-on.png')} alt="Input and Visual settings on" width="500"/>
```
@LoryApiFortress Removed

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+5. The result should look like the following:
+
+  <img src={useBaseUrl('/img/api-testing/assert-exists-final.png')} alt="What the Assert request should look like" width="500"/>
```
@LoryApiFortress Changed to "payload.downloads"

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+      <img src={useBaseUrl('/img/api-testing/post-body-window.png')} alt="The Post body window"/>
+
+4. Click the click the **Confirm changes** icon and proceed with the test.
```
@LoryApiFortress Fixed, thanks for catching that.

lauramthomas commented on docs/api-testing/composer.md:
````diff
@@ -2,82 +2,601 @@
+  ${D.format(D.plusDays(D.nowMillis(),35), 'yyyy-MM-DD','America/New_York')}
+  ```
+
+### Convert a Timestamp to Milliseconds
````
@LoryApiFortress Fixed

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+## Integration Tests
+An integration test is a test in which you examine a complete flow of calls, simulating what an API user would experience.
+
+Integration testing is critical to creating a strong API testing strategy. Microservices are built to work together, and an integration test allows you to create end-to-end tests that resemble common user flows. While only testing individual endpoints is a good start, this method will miss a large number of problems that occur when all services need to work together.
```
@LoryApiFortress I changed the "Integration testing is critical..." paragraph to "Integration testing is critical for creating a strong API testing strategy. An integration test allows you to create end-to-end tests that resemble common user flows. While only testing individual endpoints is a good start, this method will miss a large number of problems that occur when all services need to work together." 

Are the other three paragraph okay or should they be removed?

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+Integration testing is critical to creating a strong API testing strategy. Microservices are built to work together, and an integration test allows you to create end-to-end tests that resemble common user flows. While only testing individual endpoints is a good start, this method will miss a large number of problems that occur when all services need to work together.
+
+It is imperative to not just exercise endpoints, but validate that an entire series of microservices are working. It’s best to do that by writing tests that emulate common and uncommon user flows. A critical part of that work involves creating reusable variables to allow the test to work at any time, with any data.
```
@LoryApiFortress I removed all except the updated "Integration testing is critical..." paragraph, as in my last comment.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+It is imperative to not just exercise endpoints, but validate that an entire series of microservices are working. It’s best to do that by writing tests that emulate common and uncommon user flows. A critical part of that work involves creating reusable variables to allow the test to work at any time, with any data.
+
+By making a request for a fresh token at the beginning of the sequence, and then assigning it to a variable, you will know that any time you run this test, you’re doing so with a valid access token, which is automatically being passed to all follow-up calls.
```
@LoryApiFortress See my last comment

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+3. Make follow-up calls.
+
+  In the following example, the API has a cart function that requires a user token to add items to a cart or view items currently in the cart. Use a `PUT` request to the cart endpoint to update the cart. Use the access token granted by the authentication server to add items to a cart by setting the `token` header to `${access_token}`.
```
@LoryApiFortress Here "example" refers to the screenshot, not a different scenario. I can change it to "In the following screenshot..." but I don't think the original is misleading.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  1. Call the product listing endpoint and assign the response to the `productsPayload` variable.
+
+  2. Add a `for each` assertion and reference the `productsPayload.products` object.
```
@LoryApiFortress Fixed

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  4. Add a new `Set (variable)` assertion to set the `id` variable as every single `productsPayload.product` that is returned. In the following example, the string is `${_1.id}`. The system uses `_1` automatically when recognizing a subroutine, which makes it easier when there are multiple sub-levels.
+
+    <img src={useBaseUrl('/img/api-testing/int-test-endpoints.png')} alt="Testing interactions between endpoints" width="600"/>
```
@LoryApiFortress I originally replaced it with a dummy url because yours includes "apifortress," which will be removed as part of rebranding. In addition, screenshots are meant to orient the user, not give them values to enter as they work through steps. If the user is meant to enter a value, including it in a screenshot is a bad way to do that because they can't select and copy anything.

However, I removed my dummy url from other screenshots and reverted to the apifortress url because making all the screenshots consistent would take too much time. I must have missed this one, and I will fix it. At some point, after the UI is rebranded, I **will** have to go back through all the docs and update the screenshots.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+* `fetch` is the total download time of the payload.
+* `overall` is fetch and latency combined.
+
+The following is the same example, but in **Visual** view:
```
@LoryApiFortress I recreated the screenshot you included in your how-to guide. If it should be different, please give me specifics.

lauramthomas replied:
@LoryApiFortress Okay. In your how-to guide you wrote "Here is what it looks like in the VISUAL view using the Assert-Less component." Is that not the same?

Can you please give me specifics when you provide feedback like this? Should I change the sentence or do we need to add additional examples? Or something else?

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
 
 ### Visual View and Code View
-This button toggles back and forth between the Visual and Code Test Composer views. You can make calls and add assertions for testing your APIs, and insert variables wherever needed. You can use either, depending on which you're more comfortable with.
+This button toggles between the Visual and Code views in the Composer. You can make calls and add assertions for testing your APIs, and insert variables wherever needed. You can use either, depending on which you're more comfortable with.
```
@LoryApiFortress I will re-write this, but note that that is how it was originally written.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+1. In Sauce Labs, click **API Testing**.
 
-To go back and edit a Test at any time, go into your Project's **Tests** tab, hover over a test name, then click **Edit Test** (pencil icon).<br/><img src={useBaseUrl('img/api-fortress/2020/09/editTest.png')} alt="Components available" width="600" />
+  <img src={useBaseUrl('/img/api-testing/api-testing-nav.png')} alt="Navigating to API Testing" width="400"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
-:::tip
-For detailed information, see [Request (I/O) Components](/api-testing/composer/io-components/).
-:::
+  <img src={useBaseUrl('/img/api-testing/test-create-from-scratch-nav.png')} alt="Navigating to the New Test window" width="450"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
-15. Then select the **Run** icon <img src={useBaseUrl('img/api-fortress/2021/01/run.png')} alt="run test icon"/> directly next to the Save icon.<img src={useBaseUrl('img/api-fortress/2021/01/composerToolbar.png')} alt="Test Composer Tool Bar"/>
+For more information, see [Input Sets](/api-testing/composer/#input-sets) and [Visual View and Code View](/api-testing/composer/#visual-view-and-code-view).
+
+### Edit a Test
```
@LoryApiFortress I don't know what you mean exactly. Do you mean that "Edit a Test" should be at the end of the "Creating a Test with the Composer" section?

lauramthomas replied:
@LoryApiFortress I disagree. The doc says "To edit a test at any time...," which implies any time after creating the test, which is the section right before Edit a Test. I will leave it where it is.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+### Edit a Test
+To edit a test at any time, on the **Projects** page, on the **Tests** tab, hover over a test name and then click **Edit Test**.
+
+<img src={useBaseUrl('/img/api-testing/edit-test-nav.png')} alt="Navigating to the test editor" width="300"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+To create a simple `GET` request and validate that response is correct:
+  1. In API Testing, on the **Compose** page, click the **Add component** button.
+
+  <img src={useBaseUrl('/img/api-testing/add-component-nav.png')} alt="Navigating to the Add component screen" width="600"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  2. In the list of component options, click the **GET** component.
+
+  <img src={useBaseUrl('/img/api-testing/get-request-nav.png')} alt="Navigating to the GET request window" width="400"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+  3. In the **GET request** window, in the **Url** field, enter **ht<span>tps://</span>api.us-west-1.saucelabs.com/rest/v1/public/tunnels/info/versions**.
+
+  This URL will return a json response body.
+  4. In the **Variable** field, enter **payload**. This variable stores the response.
```
@LoryApiFortress Changed

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+
+  5. Leave the rest of the fields blank and then click the **Confirm changes** icon.
+
+  <img src={useBaseUrl('/img/api-testing/get-request-confirm-changes.png')} alt="Confirm changes icon" width="200"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+### Run the Test
+In the Composer, click the **Save** icon and then click **Run**.
+
+<img src={useBaseUrl('/img/api-testing/run-test-save-run.png')} alt="Save and Run icons in the Composer" width="500"/>
```
@LoryApiFortress If you want to have a discussion offline about screenshot standards and guidelines, please set something up. Otherwise, I ask that you trust me to do my job.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
+There are several ways you can compose a request body in Sauce Labs API Testing, ranging from simple to complex.
+
+:::note
+The included examples use the **POST** method, but all examples can be applied to other methods. In addition, the example scenarios use request bodies, but can be used with headers or parameters.
```
@LoryApiFortress Removed

lauramthomas commented on docs/api-testing/composer.md:
````diff
@@ -2,82 +2,601 @@
+  ```  
+
+3. Create the new data structure by adding a `SET` component.
+
````
@LoryApiFortress Fixed

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
 
 
-## How to Write a Composer Test
+## Creating a Test with the Composer
```
@LoryApiFortress But the user is, in fact, creating a test in the Composer, so it is a valid title.

lauramthomas commented on docs/api-testing/composer.md:
```diff
@@ -2,82 +2,601 @@
 
-#### Add I/O Request Component
-Create a simple `GET` request and validate that response is correct.
+  <img src={useBaseUrl('/img/api-testing/composer-nav.png')} alt="Navigating to the Composer" width="700"/>
```
@LoryApiFortress Which comment above?

lauramthomas replied:
@LoryApiFortress Okay, I think I addressed this with the previous comment.

lauramthomas commented:
@LoryApiFortress I believe I addressed all of your comments. If I don't receive answers to my open questions by tomorrow (Wednesday) morning I will merge as-is.

LoryApiFortress reviewed (commented):
@lauramthomas I hope I answered all your questions, before merging can I see the document with all the changes (right now, I only see your comment as "fixed" but I don't see the changes)

lauramthomas commented:
@LoryApiFortress All of my changes are in the api-composer branch. Are you not able to see them?
